exonum/src/runtime/dispatcher/schema.rs

```diff
@@ -149,17 +149,54 @@ impl Schema<&Fork> {
             return Err(Error::ServiceIdExists);
         }
 
-        let id = spec.id;
-        let name = spec.name.clone();
+        let instance_id = spec.id;
+        let instance_name = spec.name.clone();
+        let pending_status = InstanceStatus::Active;
+
         instances.put(
-            &name,
+            &instance_name,
             InstanceState {
-                spec: spec.clone(),
-                status: InstanceStatus::Pending,
+                spec,
+                status: None,
+                pending_status: Some(pending_status),
             },
         );
-        instance_ids.put(&id, name);
-        self.pending_instances().push(spec);
+        self.pending_instances().put(&instance_name, pending_status);
+        instance_ids.put(&instance_id, instance_name);
+        Ok(())
+    }
+
+    /// Adds information about stopping service instance to the schema.
+    pub(crate) fn initiate_stopping_service(
+        &mut self,
+        instance_id: InstanceId,
+    ) -> Result<(), Error> {
+        let mut instances = self.instances();
+        let mut pending_instances = self.pending_instances();
+
+        let instance_name = self
+            .instance_ids()
+            .get(&instance_id)
+            .ok_or(Error::IncorrectInstanceId)?;
+
+        let mut state = instances
+            .get(&instance_name)
+            .expect("BUG: Instance identifier exists but the corresponding instance is missing.");
+
+        match state.status {
+            Some(InstanceStatus::Active) => {}
+            _ => return Err(Error::ServiceNotActive),
+        }
+
+        if state.pending_status.is_some() {
+            return Err(Error::ServicePending);
+        }
+
+        // Modify instance status.
+        let pending_status = InstanceStatus::Stopped;
+        state.pending_status = Some(pending_status);
+        pending_instances.put(&instance_name, pending_status);
+        instances.put(&instance_name, state);
```

Comment:
Hmm... It looks a bit odd that we add `Stopped` service as pending (which means that this action is to be done yet), and update the instance state with the same value (which means that service is already stopped).
Wouldn't it cause the weird behaviour when service is not stopped yet but considered so by the core?